networkx/algorithms/traversal/breadth_first_search.py

```diff
@@ -426,6 +428,105 @@ def bfs_layers(G, sources):
         current_layer = next_layer
 
 
+REVERSE_EDGE = "reverse"
+TREE_EDGE = "tree"
+FORWARD_EDGE = "forward"
+LEVEL_EDGE = "level"
+
+
+def bfs_labeled_edges(G, sources):
+    """Iterate over edges in a breadth-first search (BFS) labeled by type.
+
+    We generate triple of the form (*u*, *v*, *d*), where (*u*, *v*) is the
+    edge being explored in the breadth-first search and *d* is one of the
+    strings 'tree', 'forward', 'level', or 'reverse'.  A 'tree' edge is one in
+    which *v* is first discovered and placed into the layer below *u*.  A
+    'forward' edge is one in which *u* is on the layer above *v* and *v* has
+    already been discovered.  A 'level' edge is one in which both *u* and *v*
+    occur on the same layer.  A 'reverse' edge is one in which *u* is on a layer
+    below *v*.
+
+    We emit each edge exactly once.  In an undirected graph, 'reverse' edges do
+    not occur, because each is discovered either as a 'tree' or 'forward' edge.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        A graph over which to find the layers using breadth-first search.
+
+    sources : node in `G` or list of nodes in `G`
+        Starting nodes for single source or multiple sources breadth-first search
+
+    Yields
+    ------
+    edges: generator
+       A generator of triples (*u*, *v*, *d*) where (*u*, *v*) is the edge being
+       explored and *d* is described above.
+
+    Examples
+    --------
+    >>> G = nx.cycle_graph(4, create_using = nx.DiGraph)
+    >>> list(nx.bfs_labeled_edges(G, 0))
+    [(0, 1, 'tree'), (1, 2, 'tree'), (2, 3, 'tree'), (3, 0, 'reverse')]
+    >>> G = nx.complete_graph(3)
+    >>> list(nx.bfs_labeled_edges(G, 0))
+    [(0, 1, 'tree'), (0, 2, 'tree'), (2, 1, 'level')]
+    >>> list(nx.bfs_labeled_edges(G, [0, 1]))
+    [(0, 2, 'tree'), (1, 0, 'level'), (1, 2, 'forward')]
+    """
+    if G.is_directed():
+        yield from _directed_bfs_labeled_edges(G, sources)
+    else:
+        yield from _undirected_bfs_labeled_edges(G, sources)
+
+
+def _undirected_bfs_labeled_edges(G, sources):
+    if sources in G:
+        sources = [sources]
+
+    depth = {s: 0 for s in sources}
+    visited = set()
+    queue = deque(depth.items())
+    while queue:
+        u, du = queue.popleft()
+        visited.add(u)
+        for v in G[u]:
+            dv = depth.get(v)
+            if dv is None:
+                depth[v] = dv = du + 1
+                queue.append((v, dv))
+                yield u, v, TREE_EDGE
+            else:
+                if du == dv:
+                    # could use either "in" or "not in" -- this emits self-loops
```

Comment:
I'm not sure how to read the comment: **could use either "in" or "not in"**
I'm guessing it means that we could choose to report self-loops or not.

It seems like we should report them, so every edges is reported once.
Is there a reason not to report self-loops?

We could also create another edge type: LOOP_EDGE

Reply:
I wrote it in such a way that self-loops are emitted -- "in visited" emits self-loops, "not in visited" omits them :)

Reply:
Sorry, that verbiage is obtuse.  Self-loops are level edges... while they can produce false-positives for girth cycle candidates, that will eventually be corrected when the girth computation starts at a node with a self-loop.  We could check for self-loops and bail early with an answer of 1, or let it be.  My feeling is that if a downstream consumer of this function wants to handle self-loops, it's inexpensive to do so.